README.md
# flutter_titled_container

Display a title on top of the Container.

The widget works best with a child container : Container

You can set the following properties :

- title           - The text to display 
- titleColor      - The color (foreground) fors title, default is black
- fontSize        - The font size for title, default is 14.0
- backgroundColor - The background color for title, default is white
- textAlign       - The alignment for text. i.e. Lef, Center or Right

See the screenshot

![](https://github.com/johandb/flutter_titled_container/blob/main/supplement/app.png)
![](https://github.com/johandb/flutter_titled_container/blob/main/supplement/app-r.png)
![](https://github.com/johandb/flutter_titled_container/blob/main/supplement/app-l.png)

## Example

```dart
import 'package:flutter/material.dart';
import 'package:flutter_titled_container/flutter_titled_container.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      appBar: AppBar(
        centerTitle: true,
        title: Text('Titled Container'),
      ),
      body: Center(
        child: TitledContainer(
          titleColor: Colors.blue,
          title: 'Container Title',
          textAlign: TextAlignTitledContainer.Center,
          fontSize: 16.0,
          backgroundColor: Colors.white,
          child: Container(
            width: 250.0,
            height: 200.0,
            decoration: BoxDecoration(
              border: Border.all(
                color: Colors.blue,
              ),
              borderRadius: BorderRadius.all(
                Radius.circular(10.0),
              ),
            ),
            child: Center(
              child: Text(
                'Some text',
                style: TextStyle(fontSize: 28.0),
              ),
            ),
          ),
        ),
      ),
    );
  }
}

```


CHANGELOG.md
## [1.0.7] - May 31, 2021

* Removed prints

## [1.0.6] - May 31, 2021

* Added text align for title. You can now align text Left, Center or Rigth

## [1.0.5] - April 19, 2021

* Updated the description in pubspec.yml

## [1.0.4] - April 15, 2021

* Updated the description in pubspec.yml

## [1.0.3] - April 15, 2021

* Added property to set the background color for title
* Added default values
  - titleColor      - If this property is omitted it defaults to black
  - backgroundColor - If this property is omitted it defaults to white
  - fontSize        - If this property is omitted it default to 14.0sss 

## [1.0.2] - April 14, 2021

* Updated the screenshot.
* Set background color to white in example. 

## [1.0.1] - April 14, 2021

* Updated the README

## [1.0.0] - April 14, 2021

* Initial release
* TitledContainer displays a title on the top of a Container widget. The Container must be
  a child of the TitledContainer widget.
* You can set the color, font-size and the title to be displayed
* This widget uses the childs constraints.



test/flutter_titled_container_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_titled_container/flutter_titled_container.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/flutter_titled_container.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter/widgets.dart';

enum TextAlignTitledContainer { Left, Center, Right }

class TitledContainer extends SingleChildRenderObjectWidget {
  const TitledContainer({
    Key? key,
    required Widget child,
    titleColor,
    required this.title,
    textAlign,
    fontSize,
    backgroundColor,
  })  : fontSize = fontSize ?? 14.0,
        titleColor = titleColor ?? const Color.fromRGBO(0, 0, 0, 1.0),
        textAlign = textAlign ?? TextAlignTitledContainer.Left,
        backgroundColor = backgroundColor ?? const Color.fromRGBO(255, 255, 255, 1.0),
        super(key: key, child: child);

  final Color titleColor;
  final Color backgroundColor;
  final String title;
  final TextAlignTitledContainer textAlign;
  final double fontSize;

  @override
  RenderObject createRenderObject(BuildContext context) {
    return RenderTitledContainer(
      titleColor: titleColor,
      title: title,
      fontSize: fontSize,
      backgroundColor: backgroundColor,
      textAlign: textAlign,
    );
  }

  @override
  void updateRenderObject(BuildContext context, RenderTitledContainer renderObject) {
    renderObject..titleColor = titleColor;
    renderObject..backgroundColor = backgroundColor;
    renderObject..title = title;
    renderObject..fontSize = fontSize;
    renderObject..textAlign = textAlign;
  }
}

class RenderTitledContainer extends RenderBox with RenderObjectWithChildMixin<RenderBox> {
  RenderTitledContainer({
    required Color titleColor,
    required String title,
    required double fontSize,
    required Color backgroundColor,
    required TextAlignTitledContainer textAlign,
  })   : _titleColor = titleColor,
        _title = title,
        _textAlign = textAlign,
        _backgroundColor = backgroundColor,
        _fontSize = fontSize;

  Color get titleColor => _titleColor;
  Color _titleColor;
  set titleColor(Color value) {
    if (_titleColor == value) return;
    _titleColor = value;
    markNeedsPaint();
  }

  Color get backgroundColor => _backgroundColor;
  Color _backgroundColor;
  set backgroundColor(Color value) {
    if (_backgroundColor == value) return;
    _backgroundColor = value;
    markNeedsPaint();
  }

  String get title => _title;
  String _title;
  set title(String value) {
    if (_title == value) return;
    _title = value;
    markNeedsPaint();
  }

  TextAlignTitledContainer get textAlign => _textAlign;
  TextAlignTitledContainer _textAlign;
  set textAlign(TextAlignTitledContainer value) {
    if (_textAlign == value) return;
    _textAlign = value;
    markNeedsPaint();
  }

  double get fontSize => _fontSize;
  double _fontSize;
  set fontSize(double value) {
    if (_fontSize == value) return;
    _fontSize = value;
    markNeedsPaint();
  }

  @override
  void performLayout() {
    if (child == null) {
      size = constraints.smallest;
    } else {
      child!.layout(constraints, parentUsesSize: true);
      size = constraints.constrain(Size(child!.size.width, child!.size.height));
    }
  }

  @override
  void paint(PaintingContext context, Offset offset) {
    if (child != null) {
      context.paintChild(child!, offset);

      final canvas = context.canvas;
      canvas.save();
      canvas.translate(offset.dx, offset.dy);
      final textSpan = TextSpan(
        text: ' $title ',
        style: TextStyle(
          color: titleColor,
          fontSize: fontSize,
          height: 1.0,
          backgroundColor: backgroundColor,
        ),
      );
      final textPainter = TextPainter(
        text: textSpan,
        textDirection: TextDirection.ltr,
      );
      textPainter.layout(
        minWidth: 0,
        maxWidth: size.width,
      );

      final TextPainter txtPainter = TextPainter(text: textSpan, textDirection: TextDirection.ltr);
      txtPainter.layout(minWidth: 0, maxWidth: double.infinity);
      double xPos = 10.0;
      switch (textAlign) {
        case TextAlignTitledContainer.Center:
          xPos = (size.width - txtPainter.size.width) / 2.0;
          break;
        case TextAlignTitledContainer.Right:
          xPos = (size.width - txtPainter.size.width - 10);
          break;
        default:
          xPos = 10.0;
      }
      final titleOffset = Offset(xPos, -fontSize / 2);
      textPainter.paint(canvas, titleOffset);

      canvas.restore();
    }
  }

  @protected
  bool get alwaysNeedsCompositing => true;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_titled_container/flutter_titled_container.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      appBar: AppBar(
        centerTitle: true,
        title: Text('Titled Container'),
      ),
      body: Center(
        child: TitledContainer(
          titleColor: Colors.blue,
          title: 'Center Title',
          textAlign: TextAlignTitledContainer.Left,
          fontSize: 14.0,
          backgroundColor: Colors.white,
          child: Container(
            width: 250.0,
            height: 200.0,
            decoration: BoxDecoration(
              color: Colors.white,
              border: Border.all(
                color: Colors.blue,
              ),
              borderRadius: BorderRadius.all(
                Radius.circular(10.0),
              ),
            ),
            child: Center(
              child: Text(
                'Center Align text',
                style: TextStyle(fontSize: 28.0),
              ),
            ),
          ),
        ),
      ),
    );
  }
}
